### ISSUE TITLE
Issue using `async_trait` with `zigbuild` for x86_64-unknown-linux-gnu

### ISSUE THREAD

**Issue by nmoutschen** [messense/cargo-zigbuild#4]
Hi! 👋 First of all, thanks a lot for this awesome project!

I'm trying to use `cargo zigbuild` with [this project](https://github.com/aws-samples/serverless-rust-demo) that uses the [`async_trait` crate](https://crates.io/crates/async-trait) from Ubuntu on x86_64.

When I use `cargo zigbuild --target aarch64-unknown-linux-gnu`, it works without any error. Same with `cargo build --target x86_64-unknown-linux-gnu`.
However, when I run `cargo zigbuild --target x86_64-unknown-linux-gnu`, I get a pretty lengthy error about async functions not supported in traits, and that thiserror cannot allocate memory in static TLS block.

Error trace from compilation:

```
error[E0706]: functions in traits cannot be declared `async`
  --> src/event_bus/eventbridge/mod.rs:32:5
   |
32 |     #[instrument(skip(self))]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
33 |     async fn send_event(&self, event: &Self::E) -> Result<(), Error> {
   |     ----- `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
   = note: this error originates in the attribute macro `instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0706]: functions in traits cannot be declared `async`
  --> src/event_bus/eventbridge/mod.rs:45:5
   |
45 |     #[instrument(skip(self, events))]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
46 |     async fn send_events(&self, events: &[Self::E]) -> Result<(), Error> {
   |     ----- `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
   = note: this error originates in the attribute macro `instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0706]: functions in traits cannot be declared `async`
  --> src/event_bus/void.rs:18:5
   |
18 |       async fn send_event(&self, _: &Self::E) -> Result<(), Error> {
   |       ^----
   |       |
   |  _____`async` because of this
   | |
19 | |         Err(Error::InternalError("send_event is not supported"))
20 | |     }
   | |_____^
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/event_bus/void.rs:22:5
   |
22 |       async fn send_events(&self, _: &[Self::E]) -> Result<(), Error> {
   |       ^----
   |       |
   |  _____`async` because of this
   | |
23 | |         Err(Error::InternalError("send_events is not supported"))
24 | |     }
   | |_____^
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/event_bus/mod.rs:14:5
   |
14 |     async fn send_event(&self, event: &Self::E) -> Result<(), Error>;
   |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/event_bus/mod.rs:15:5
   |
15 |     async fn send_events(&self, events: &[Self::E]) -> Result<(), Error>;
   |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/dynamodb/mod.rs:32:5
   |
32 |     #[instrument(skip(self))]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
33 |     async fn all(&self, next: Option<&str>) -> Result<ProductRange, Error> {
   |     ----- `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
   = note: this error originates in the attribute macro `instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/dynamodb/mod.rs:60:5
   |
60 |     #[instrument(skip(self))]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
61 |     async fn get(&self, id: &str) -> Result<Option<Product>, Error> {
   |     ----- `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
   = note: this error originates in the attribute macro `instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/dynamodb/mod.rs:81:5
   |
81 |     #[instrument(skip(self))]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
82 |     async fn put(&self, product: &Product) -> Result<(), Error> {
   |     ----- `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
   = note: this error originates in the attribute macro `instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/dynamodb/mod.rs:98:5
   |
98 |     #[instrument(skip(self))]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
99 |     async fn delete(&self, id: &str) -> Result<(), Error> {
   |     ----- `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
   = note: this error originates in the attribute macro `instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/memory.rs:28:5
   |
28 |       async fn all(&self, _: Option<&str>) -> Result<ProductRange, Error> {
   |       ^----
   |       |
   |  _____`async` because of this
   | |
29 | |         Ok(ProductRange {
30 | |             products: self
31 | |                 .data
...  |
38 | |         })
39 | |     }
   | |_____^
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/memory.rs:44:5
   |
44 |       async fn get(&self, id: &str) -> Result<Option<Product>, Error> {
   |       ^----
   |       |
   |  _____`async` because of this
   | |
45 | |         Ok(self.data.read().unwrap().get(id).cloned())
46 | |     }
   | |_____^
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/memory.rs:51:5
   |
51 |       async fn put(&self, product: &Product) -> Result<(), Error> {
   |       ^----
   |       |
   |  _____`async` because of this
   | |
52 | |         self.data
53 | |             .write()
54 | |             .unwrap()
55 | |             .insert(product.id.clone(), product.clone());
56 | |         Ok(())
57 | |     }
   | |_____^
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/memory.rs:62:5
   |
62 |       async fn delete(&self, id: &str) -> Result<(), Error> {
   |       ^----
   |       |
   |  _____`async` because of this
   | |
63 | |         self.data.write().unwrap().remove(id);
64 | |         Ok(())
65 | |     }
   | |_____^
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/mod.rs:22:5
   |
22 |     async fn all(&self, next: Option<&str>) -> Result<ProductRange, Error>;
   |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/mod.rs:28:5
   |
28 |     async fn get(&self, id: &str) -> Result<Option<Product>, Error>;
   |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/mod.rs:34:5
   |
34 |     async fn put(&self, product: &Product) -> Result<(), Error>;
   |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error[E0706]: functions in traits cannot be declared `async`
  --> src/store/mod.rs:40:5
   |
40 |     async fn delete(&self, id: &str) -> Result<(), Error>;
   |     -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     `async` because of this
   |
   = note: `async` trait functions are not currently supported
   = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait

error: /home/CENSORED/projects/serverless-rust-demo/target/debug/deps/libthiserror_impl-83b2547aa6b39328.so: cannot allocate memory in static TLS block
 --> src/event_bus/eventbridge/ext.rs:2:5
  |
2 | use aws_sdk_eventbridge::model::PutEventsRequestEntry;
  |     ^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0706`.
error: could not compile `products` due to 19 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

For reference, here is one of the traits that `cargo zigbuild` complains about:

```rust
#[async_trait]
pub trait EventBus {
    type E;

    async fn send_event(&self, event: &Self::E) -> Result<(), Error>;
    async fn send_events(&self, events: &[Self::E]) -> Result<(), Error>;
}
```

**Comment by messense** [comment]
I can reproduce this issue on Ubuntu 20.04, but it seems to work fine on macOS.

**Comment by messense** [comment]
With commit https://github.com/messense/cargo-zigbuild/commit/407e2e410cd301d8983370292266ec71eb00a9e3, if don't specify a glibc version it'll skip setting up zig as linker and just use system linker instead when the cross target is exactly the same as the host target. That should avoid having to use the `target-applies-to-host` option most of the time.

**Comment by messense** [comment]
With #9 which is released in v0.5.2, the only workaround is switching to nightly Rust, cargo-zigbuild will take care of enabling the `target-applies-to-host ` option automatically.
